src/libmongoc/src/mongoc/mongoc-scram.c

```diff
@@ -1052,93 +1051,185 @@ _mongoc_sasl_prep_impl (const char *name,
+      }
+   }
+   char *utf8_pre_norm =
+      (char *) bson_malloc (sizeof (char) * (utf8_pre_norm_len + 1));
+
+   char *loc = utf8_pre_norm;
+   for (size_t i = 0; i < num_chars; ++i) {
+      const ssize_t utf8_char_length =
+         _mongoc_utf8_code_point_to_str (utf8_codepoints[i], loc);
+      if (utf8_char_length == -1) {
+         bson_free (utf8_pre_norm);
+         bson_free (utf8_codepoints);
+         SASL_PREP_ERR_RETURN ("invalid Unicode code point in %s");
+      }
+      loc += utf8_char_length;
+   }
+   *loc = '\0';
+
+   out_utf8 = (uint8_t *) utf8proc_NFKC ((utf8proc_uint8_t *) utf8_pre_norm);
+
+   // the last two steps are both checks for characters that should not be
+   // allowed. Because the normalization step is guarenteed to not create any
+   // characters that will cause an error, we will use the utf8_codepoints
+   // codepoints to check (pre-normalization) as to avoid converting back and
+   // forth from UTF-8 to unicode codepoints.
+
+   // c. Prohibit -- Check for any characters
+   // that are not allowed in the output. If any are found, return an error.
+
+   for (size_t i = 0; i < num_chars; ++i) {
+      if (_mongoc_utf8_code_point_is_in_table (
+             utf8_codepoints[i],
+             prohibited_output_ranges,
+             sizeof (prohibited_output_ranges) / sizeof (uint32_t)) ||
+          _mongoc_utf8_code_point_is_in_table (
+             utf8_codepoints[i],
+             unassigned_codepoint_ranges,
+             sizeof (unassigned_codepoint_ranges) / sizeof (uint32_t))) {
+         bson_free (out_utf8);
+         bson_free (utf8_pre_norm);
+         bson_free (utf8_codepoints);
+         SASL_PREP_ERR_RETURN ("prohibited character included in %s");
+      }
+   }
+
+   // d. Check bidi -- Possibly check for right-to-left characters, and if
+   // any are found, make sure that the whole string satisfies the
+   // requirements for bidirectional strings.  If the string does not
+   // satisfy the requirements for bidirectional strings, return an
+   // error.
+
+   // note: bidi stands for directional (text). Most characters are displayed
+   // left to right but some are displayed right to left. The requirements are
+   // as follows:
+   // 1. If a string contains any RandALCat character, it can't contain an LCat
+   // character
+   // 2. If it contains an RandALCat character, there must be an RandALCat
+   // character at the beginning and the end of the string (does not have to be
+   // the same character)
+   bool contains_LCat = false;
+   bool contains_RandALCar = false;
+
+
+   for (size_t i = 0; i < num_chars; ++i) {
+      if (_mongoc_utf8_code_point_is_in_table (utf8_codepoints[i],
+                                               LCat_bidi_ranges,
+                                               sizeof (LCat_bidi_ranges) /
+                                                  sizeof (uint32_t))) {
+         contains_LCat = true;
+         if (contains_RandALCar)
+            break;
```

Comment:
```suggestion
```

Suggest removing unnecessary early break for consistency with assignment of `contains_RandALCar = true;`